### ISSUE TITLE
hlo_live_range reports assertion failure when running an HLO module file through run_hlo_module or multihost_hlo_runner/hlo_runner_main

### ISSUE THREAD

**Issue by shawnwang18** [openxla/xla#4762]
Error messages: 

2023-08-04 14:35:15.565823: F xla/hlo/utils/hlo_live_range.cc:213] Check failed: async_done_it != async_context->AsyncInstructions().end()

Reproduce commands:  

./bazel-5.3.0-linux-x86_64 build \
--verbose_failures \
-c opt \
--config=cuda \
--nocheck_visibility \
--dynamic_mode=off  \
--copt="-Wno-error=switch" \
--copt="-DLLVM_ENABLE_DUMP=ON" \
--action_env TF_CUDA_COMPUTE_CAPABILITIES=compute_80 \
//xla/tools/multihost_hlo_runner:hlo_runner_main

./bazel-bin/xla/tools/multihost_hlo_runner/hlo_runner_main \
--device_type=gpu \
--use_spmd_partitioning=true \
--num_partitions=8 \
--num_replicas=1 \
--hlo_file="module_0131.pjit__wrapped_step_fn.sm_8.0_gpu_after_optimizations.txt"

[module_0131.pjit__wrapped_step_fn.sm_8.0_gpu_after_optimizations.txt](https://github.com/openxla/xla/files/12262007/module_0131.pjit__wrapped_step_fn.sm_8.0_gpu_after_optimizations.txt)

**Comment by akuegel** [comment]
If you run HLO after optimizations (which you do in this case), I think you should turn off optimization passes (--disable_all_hlo_passes). Otherwise, pass HLO before optimizations instead.
Maybe this does not help for this case (I haven't tried), but you can definitely run into several problems if you try to run optimization passes again on an already optimized HloModule.

**Comment by shawnwang18** [comment]
Thanks @akuegel 
It runs fine for the before optimization HLO file.  But for the after optimization HLO file, I used the env: 
export XLA_FLAGS='--xla_disable_all_hlo_passes'  
to disable optimization passes for it, it still reports the assertion error.  Not sure if it is as expected.

**Comment by akuegel** [comment]
Yes, this sounds wrong. If you disable all hlo passes, it should work. So from the symptoms it sounds as if we do not dump the full state after optimization passes (theoretically running with disabling all hlo passes should just continue at the point when the dump was created).

**Comment by akuegel** [comment]
Unfortunately by now the reproducer HLO cannot be parsed anymore:

```INVALID_ARGUMENT: Syntax error when trying to parse the text as a HloModule:
was parsing 1645:93: error: unexpected attribute "channel_id". Allowed attributes: backend_config, parameter_replication, frontend_attributes, sharding, statistics, origin, control-predecessors, metadata, async_execution_thread
  reduce-scatter-done = bf16[2,8192,4096]{1,0,2} reduce-scatter-done(reduce-scatter-start), channel_id=291, replica_groups={{0,1},{2,3},{4,5},{6,7}}, use_global_device_ids=true, dimensions={2}, to_apply=add
```

@shawnwang18 Did you recently still run into a similar issue, and if yes, could you upload another reproducer?

**Comment by akuegel** [comment]
I guess we can close this.
